src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -793,6 +793,12 @@ internal void DeactivateActivationOnIdle(ActivationData data)
                         data.AddOnInactive(() => DestroyActivationVoid(data));
                     }
                 }
+                else if (data.State == ActivationState.Activating)
+                {
+                    throw new GrainRefusedToActivateException(String.Format(
```

Comment:
I think we should throw an `InvalidOperationException` instead with a message that `DeactivateOnIdle()` cannot be called before the grain is activated (from within its constructor or `OnActivateAsync()`).

I think the semantics we want is that a grain cannot refuse to activate, it can only fail to. Of course, a grain can work around it and refuse by causing a failure - by throwing an exception, but form the runtime's perspective there shouldn't be a semantic difference between those cases, and no notion of refusal to activate.